cmds/bootcm4.c

```diff
@@ -0,0 +1,179 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system loader
+ *
+ * boot CM4 CPU i.MX RT117x
+ *
+ * Copyright 2022-2023 Phoenix Systems
+ * Author: Gerard Swiderski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "cmd.h"
+
+#include <devices/devs.h>
+#include <hal/hal.h>
+#include <lib/lib.h>
+#include <phfs/phfs.h>
+
+
+#define CM4_BOOT_MEMSIZE 0x00040000u
+#define CM4_BOOT_ADDRESS 0x20200000u
+#define CM4_VTOR_ADDRESS 0x20200100u
+
+
+static unsigned int isImgLoaded = 0u;
+
+
+static void cmd_bootcm4Info(void)
+{
+	lib_printf("boots CM4 core image, use <-h> to see usage");
+}
+
+
+static void print_usage(const char *name)
+{
+	lib_printf(
+		"Usage: %s <options> [<device image>]\n"
+		"\t-b  boot CM4 core\n"
+		"\t-l  load binary image using phfs <device> and <image>",
+		name);
+}
+
+
+static int loadImagePhfs(const char *phfsDev, const char *phfsFile)
+{
+	u8 buff[SIZE_MSG_BUFF];
+	handler_t h;
+	addr_t offs = 0u;
+	addr_t addr = CM4_BOOT_ADDRESS;
+
+	int res = phfs_open(phfsDev, phfsFile, 0, &h);
+
+	if (res < 0) {
+		log_error("Can't open '%s' on '%s'", phfsFile, phfsDev);
+		return res;
+	}
+
+	do {
+		res = phfs_read(h, offs, buff, sizeof(buff));
+		if (res < 0) {
+			log_error("Can't read '%s' from '%s'", phfsFile, phfsDev);
+			break;
+		}
+
+		hal_memcpy((void *)addr, buff, res);
+		offs += res;
+		addr += res;
+
+		if (offs > CM4_BOOT_MEMSIZE) {
+			log_error("Image '%s' too big", phfsFile);
+			res = -ENOSPC;
+			break;
+		}
```

Comment:
Shouldn't it be check first and then copy?
```suggestion
		offs += res;
		if (offs > CM4_BOOT_MEMSIZE) {
			log_error("Image '%s' too big", phfsFile);
			res = -ENOSPC;
			break;
		}

		hal_memcpy((void *)addr, buff, res);
		addr += res;

```